src/nvim/screen.c

```diff
@@ -1863,64 +1920,25 @@ static void fold_line(win_T *wp, long fold_count, foldinfo_T *foldinfo, linenr_T
    *    Right-left text is put in columns 0 - number-col, normal text is put
    *    in columns number-col - window-width.
    */
-  int cells;
-  int u8c, u8cc[MAX_MCO];
   int idx;
-  int c_len;
-  char_u  *p;
-  int prev_c = 0;  // previous Arabic character
-  int prev_c1 = 0;  // first composing char for prev_c
 
   if (wp->w_p_rl) {
     idx = off;
   } else {
     idx = off + col;
   }
 
-  // Store multibyte characters in ScreenLines[] et al. correctly.
-  for (p = text; *p != NUL; ) {
-    cells = utf_ptr2cells(p);
-    c_len = utfc_ptr2len(p);
-    if (col + cells > wp->w_width - (wp->w_p_rl ? col : 0)) {
-      break;
-    }
-    u8c = utfc_ptr2char(p, u8cc);
-    if (*p < 0x80 && u8cc[0] == 0) {
-      schar_from_ascii(ScreenLines[idx], *p);
-      prev_c = u8c;
-    } else {
-      if (p_arshape && !p_tbidi && arabic_char(u8c)) {
-        // Do Arabic shaping.
-        int pc, pc1, nc;
-        int pcc[MAX_MCO];
-        int firstbyte = *p;
-
-        // The idea of what is the previous and next
-        // character depends on 'rightleft'.
-        if (wp->w_p_rl) {
-          pc = prev_c;
-          pc1 = prev_c1;
-          nc = utf_ptr2char(p + c_len);
-          prev_c1 = u8cc[0];
-        } else {
-          pc = utfc_ptr2char(p + c_len, pcc);
-          nc = prev_c;
-          pc1 = pcc[0];
-        }
-        prev_c = u8c;
+  LineState s = LINE_STATE(text);
 
-        u8c = arabic_shape(u8c, &firstbyte, &u8cc[0], pc, pc1, nc);
-      } else {
-        prev_c = u8c;
-      }
-      schar_from_cc(ScreenLines[idx], u8c, u8cc);
-    }
-    if (cells > 1) {
-      ScreenLines[idx + 1][0] = 0;
+  while (*s.p != NUL) {
+    // TODO(bfredl): cargo-culted from the existing code, obviously wrong
+    int maxcells = wp->w_width - col - (wp->w_p_rl ? col : 0);
```

Comment:
"the old Vim code" might be less confusing than "the existing code", since IIUC this is referring to lines now deleted.